lib/_stream_writable.js

```diff
@@ -515,21 +521,35 @@ function needFinish(state) {
 }
 
 function prefinish(stream, state) {
-  if (!state.prefinished) {
-    state.prefinished = true;
-    stream.emit('prefinish');
+  if (!state.prefinished && !state.finalCalled) {
+    if (typeof stream._final === 'function') {
+      state.pendingcb++;
+      state.finalCalled = true;
+      process.nextTick(() => {
```

Comment:
yeah so if that is call synchronously, then if the callback in final is called synchronously then the finish event will be called before the _flush method